<font size="5"><b>Kementerian Keuangan RI </b></font><br/>
<font size="4"><b>Badan Pendidikan dan Pelatihan Keuangan<b/></font><br/>
<font size="4"><b>Pusdiklat Keuangan Umum<b/></font><br/>
### Praktikum Data Scientist Specialist

***
### Perhatikan petunjuk pengisian berikut :

1. Tujuan praktikum ini adalah untuk mengukur kemampuan Anda dalam melakukan pengolahan data dan mengaplikasikan metode-metode machine learning.
2. Anda diperbolehkan melihat file <u>jupyter-notebook yang telah disediakan sebelumnya atau menggunakan search engine </u> untuk menyelesaikan soal-soal yang disediakan.
3. Perhatikan petunjuk pengisian pada masing-masing bagian. Tambahan pekerjaan (step) untuk pengolahan data akan memberikan nilai bonus.
4. <font color="red"><b>Waktu menyelesaikan</b></font> seluruh soal berikut adalah <font color="red"><b>60 menit</b></font>.
5. Simpan file ini menjadi dalam bentuk html, Arahkan ke menu **File** kemudian **Download as** -> **html (.html)**
6. Kirim file html tersebut ke wahyuagungs@gmail.com dan aginoffice@gmail.com
***

Tuliskan Nama dan NIP Anda :

NAMA / NIP : Ninin Sapto Hargiyanto / 198607202007011001



## Data Science Story Project
> Anda diminta oleh Bank Rakyat Bersama (BeRaBe) untuk membangun model yang akan digunakan untuk mengklasifikasikan rating kredit (credit rating) seseorang berdasarkan perilakunya.

> Anda diminta untuk membuat model klasifikasi yang paling akurat dari beberapa model yang ada, **paling sedikit memuat 3 model** antara lain yaitu **Logistic Regression, RandomForest, Support Vector Machine, Multi Layer Perceptron, dan/atau model lainnya yang pernah Anda telusuri**,. 

> Anda juga diminta untuk memberikan perbandingan model berdasarkan akurasi.

> Di samping itu, Anda juga diminta untuk melakukan penjelasan atas evaluasi model yang ada.

Terdapat tahapan pada jupyter notebook yang sudah dilengkapi secara default, namun Anda diminta untuk tetap melakukan tahapan-tahapan dasar standard *machine learning* yaitu *exploratory data analysis, data cleansing,* dan lainnya sebelum membangun model.
### Import library dan load dataset

Lakukan import library yang dibutuhkan

In [60]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

Terdapat 2 file yaitu `credit card rating.csv` dan `columns.csv`. Lakukan import kedua file tersebut dan simpan ke dalam bentuk objek `DataFrame` menggunakan fungsi `read_csv`. 

In [61]:
df1 = pd.read_csv('credit card rating.csv')

In [62]:
df2 = pd.read_csv('columns.csv')

----

## I. Exploratory Data Analysis

Pada tahapan ini Anda diminta untuk melakukan eksplorasi terhadap dataset ini. Berikut ini adalah tahapan-tahapan yang harus dijalankan.

<b> 1. a. Tampilkan konten dari kedua dataframe tersebut. Apakah terdapat anomali? Menurut Anda apa yang harus kita lakukan? Jelaskan secara singkat! </b>

In [63]:
df1.head(10)

0  67    male  2   own  Unnamed: 5    little  1169   6  \
0   1  22  female  2   own      little  moderate  5951  48   
1   2  49    male  1   own      little       NaN  2096  12   
2   3  45    male  2  free      little    little  7882  42   
3   4  53    male  2  free      little    little  4870  24   
4   5  35    male  1  free         NaN       NaN  9055  36   
5   6  53    male  2   own  quite rich       NaN  2835  24   
6   7  35    male  3  rent      little  moderate  6948  36   
7   8  61    male  1   own        rich       NaN  3059  12   
8   9  28    male  3   own      little  moderate  5234  30   
9  10  25  female  2  rent      little  moderate  1295  12   

              radio/TV  good  
0             radio/TV     -  
1            education  good  
2  furniture/equipment  good  
3                  car     -  
4            education  good  
5  furniture/equipment  good  
6                  car  good  
7             radio/TV  good  
8                  car     -  
9                  car     -

In [64]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 11 columns):
0             999 non-null int64
67            999 non-null int64
male          999 non-null object
2             999 non-null int64
own           999 non-null object
Unnamed: 5    817 non-null object
little        605 non-null object
1169          999 non-null int64
6             999 non-null int64
radio/TV      999 non-null object
good          999 non-null object
dtypes: int64(5), object(6)
memory usage: 86.0+ KB


In [65]:
df2.head(10)

Empty DataFrame
Columns: [age, sex, job, housing, saving_accounts, checking_account, credit_amount, duration, purpose, customer_type]
Index: []

df1 adalah tabelnya, df2 adalah nama kolomnya, untuk df1 analisis sekilas adalah sbb:
<p>Baris pertama bukan judul tapi baris, mesti dibuat nama kolom baru. </p>
<p>Kolom no 1 0 dan isi dari kolomnya harus diremove, hasil generate auto increment database ini tidak memberikan insight dalam proses analisis. </p>
<p>Kolom no 2 berjudul 67 mesti diganti, melihat isi datanya sepertinya ini bisa diganti menjadi age</p>
<p>Kolom no 3 berjudul 2 sepertinya tipe data ordinal, kita lihat dan perdalam analisis selanjutnya</p>
<p>Kolom no 4 yang berjudul male juga mesti diganti jadi sex, isinya bisa dihot encoding menjadi 0 dan 1</p>
<p>Kolom no 5 bisa diganti menjadi ownership namanya, tipe data ordinal, bisa di-encode menjadi 1,2,3 dst</p>
<p>Kolom no 6 berisi little ganti menjadi type, isinya banyak NaN perlu di-input dengan median/mean atau diremove sekalian, tergantung hasil analisis berikutnya </p>
<p>Kolom no 7 berjudul 1169 sekilas sepertinya tidak bermakna, tapi tergantung hasil analisa selanjutnya, </p>
<p>Kolom no 8 radio/TV bisa direname menjadi media, atau diparse menjadi beberapa kolom boolean</p>
<p>Kolom no 9 berisi good ini sepertinya adalah target variable, ini ada nilai - perlu kita prediksi nilai - nya.</p>

<b> 1. b. Lakukan tindakan yang perlu harus Anda lakukan untuk mengatasi hal tersebut di bawah ini </b>

In [66]:
# import ulang, tanpa header bisa lebih mudah
# hapus dulu kolom pertama
del df1['0']
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 10 columns):
67            999 non-null int64
male          999 non-null object
2             999 non-null int64
own           999 non-null object
Unnamed: 5    817 non-null object
little        605 non-null object
1169          999 non-null int64
6             999 non-null int64
radio/TV      999 non-null object
good          999 non-null object
dtypes: int64(4), object(6)
memory usage: 78.2+ KB


In [67]:
judul = list(df2.columns.values)
judul

['age',
 'sex',
 'job',
 'housing',
 'saving_accounts',
 'checking_account',
 'credit_amount',
 'duration',
 'purpose',
 'customer_type']

In [68]:
df1 = pd.read_csv('credit card rating.csv', header=None, names =judul)
df1.head(10)

age     sex  job housing saving_accounts checking_account  credit_amount  \
0   67    male    2     own             NaN           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little              NaN           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   
5   35    male    1    free             NaN              NaN           9055   
6   53    male    2     own      quite rich              NaN           2835   
7   35    male    3    rent          little         moderate           6948   
8   61    male    1     own            rich              NaN           3059   
9   28    male    3     own          little         moderate           5234   

   duration              purpose customer_type  
0         6             radio/TV          good  
1        48             radio/TV             -  
2        12            education          good  
3        42  furniture/equipment          good  
4        24                  car             -  
5        36            education          good  
6        24  furniture/equipment          good  
7        36                  car          good  
8        12             radio/TV          good  
9        30                  car             -

<b> 1. c. Lakukan observasi terhadap data tersebut, dan jelaskan secara singkat apa yang menjadi perhatian. Apabila terdapat data `NaN` tampilkan jumlah data NaN tersebut</b>

In [69]:
for c in judul:
    print(c,':',df1[c].unique())

age : [67 22 49 45 53 35 61 28 25 24 60 32 44 31 48 26 36 39 42 34 63 27 30 57
 33 37 58 23 29 52 50 46 51 41 40 66 47 56 54 20 21 38 70 65 74 68 43 55
 64 75 19 62 59]
sex : ['male' 'female']
job : [2 1 3 0]
housing : ['own' 'free' 'rent']
saving_accounts : [nan 'little' 'quite rich' 'rich' 'moderate']
checking_account : ['little' 'moderate' nan 'rich']
credit_amount : [ 1169  5951  2096  7882  4870  9055  2835  6948  3059  5234  1295  4308
  1567  1199  1403  1282  2424  8072 12579  3430  2134  2647  2241  1804
  2069  1374   426   409  2415  6836  1913  4020  5866  1264  1474  4746
  6110  2100  1225   458  2333  1158  6204  6187  6143  1393  2299  1352
  7228  2073  5965  1262  3378  2225   783  6468  9566  1961  6229  1391
  1537  1953 14421  3181  5190  2171  1007  1819  2394  8133   730  1164
  5954  1977  1526  3965  4771  9436  3832  5943  1213  1568  1755  2315
  1412 12612  2249  1108   618  1409   797  3617  1318 15945  2012  2622
  2337  7057  1469  2323   932  1919  2445 

In [70]:
df1.describe()

age          job  credit_amount     duration
count  1000.000000  1000.000000    1000.000000  1000.000000
mean     35.546000     1.904000    3271.258000    20.903000
std      11.375469     0.653614    2822.736876    12.058814
min      19.000000     0.000000     250.000000     4.000000
25%      27.000000     2.000000    1365.500000    12.000000
50%      33.000000     2.000000    2319.500000    18.000000
75%      42.000000     2.000000    3972.250000    24.000000
max      75.000000     3.000000   18424.000000    72.000000


Anda diminta untuk melakukan cek atas data **NULL** <br>
Data **NULL** tersebut bisa berbentuk tanda `-`, spasi, `NA` ataupun tanda lainnya. Strategi atas data **NULL** tersebut bisa kita lakukan *impute* atau pengisian, atau bisa dihilangkan seluruh baris (row) data tersebut. 

Data anomali lainnya, bisa berupa **kesalahan ketik (typo)** ataupun nilai yang sebenarnya sama tetapi dituliskan berbeda.

<b> 1. d. Apakah semua nilai di dalam kolom tersebut sudah bisa langsung digunakan? Temukan nilai yang seharusnya tidak boleh berada di situ</b>

In [71]:
# saving_accounts ada NaN

In [72]:
# checking_account ada NaN

----

## II. Data Pre-processing
Tahapan lanjutan adalah melakukan pembersihan atau organising atas data yang dimiliki.

> Namun, diketahui bahwa, terdapat *feature* (kolom) terdapat nilai `NaN`, di samping nilai yang sesungguhnya.


> Di samping itu, terdapat beberapa karakter yang tidak sesuai dengan yang seharusnya

**2. Langkah apa yang akan Anda lakukan?**<br/>

Tuliskan Langkah-langkah Anda di bawah ini:

In [73]:
# Merubah sex dari female dan male menjai 0 dan 1

In [74]:
# Merubah housing menjadi ordinal data free = 0, rent = 1, own = 2

<b> 3. Lakukan langkah Yang Anda telah jelaskan di atas </b>

In [75]:
df1['sex'].unique

<bound method Series.unique of 0        male
1      female
2        male
3        male
4        male
        ...  
995    female
996      male
997      male
998      male
999      male
Name: sex, Length: 1000, dtype: object>

In [76]:
df1['sex'] = df1['sex'].map({'male':1,'female':0}).astype(int)

In [77]:
df1['housing'].unique()

array(['own', 'free', 'rent'], dtype=object)

In [78]:
def changehousing(x):
    if x =='free':
        return 0
    elif x =='rent':
        return 1
    elif x =='own':
        return 2
    else:
        return x

In [79]:
df1.housing = df1.housing.apply(changehousing)
df1['housing'].unique()

array([2, 0, 1])

In [80]:
# Hapus record yang memiliki nilai NaN pada attributnya
df1 = df1.dropna(how='any')

## III. Analysis
Tahapan sebelum dilakukan pembangunan model, Lakukan Analysis atas data tersebut di atas. <br>
Dalam melakukan pembangunan model, Anda dihadapkan oleh banyak kolom yang tersedia dimana beberapa kolom merupakan data categorical

**4.a. Apa yang akan Anda lakukan?**

Lakukan Encoding

<b> 4. b. **Lakukan langkah Anda di atas** </b>

In [81]:
df1['saving_accounts'].unique()

array(['little', 'moderate', 'quite rich', 'rich'], dtype=object)

In [82]:
df1['checking_account'].unique()

array(['moderate', 'little', 'rich'], dtype=object)

In [83]:
def changeaccount(x):
    if x =='little':
        return 1
    elif x =='moderate':
        return 2
    elif x =='quite rich':
        return 3
    elif x =='rich':
        return 4
    else:
        return x

In [84]:
df1.saving_accounts = df1.saving_accounts.apply(changeaccount)
df1['saving_accounts'].unique()

array([1, 2, 3, 4])

In [85]:
df1.checking_account = df1.checking_account.apply(changeaccount)
df1['checking_account'].unique()

array([2, 1, 4])

In [86]:
df1.columns

Index(['age', 'sex', 'job', 'housing', 'saving_accounts', 'checking_account',
       'credit_amount', 'duration', 'purpose', 'customer_type'],
      dtype='object')

In [87]:
from sklearn.preprocessing import LabelEncoder

df1.purpose = LabelEncoder().fit_transform(df1.purpose)

In [88]:
df1.head()

age  sex  job  housing  saving_accounts  checking_account  credit_amount  \
1   22    0    2        2                1                 2           5951   
3   45    1    2        0                1                 1           7882   
4   53    1    2        0                1                 1           4870   
7   35    1    3        1                1                 2           6948   
9   28    1    3        2                1                 2           5234   

   duration  purpose customer_type  
1        48        5             -  
3        42        4          good  
4        24        1             -  
7        36        1          good  
9        30        1             -

Diasumsikan bahwa data yang akan kita gunakan telah melalui serangkaian pembersihan data dan seterusnya. Tugas selanjutnya adalah membuat model untuk memprediksi customer_type berdasarkan atribut lainnya.

### Lakukan train test split
Sebelum membangun model, langkah yang harus dilakukan adalah memecah target variabel dan independent variabel. 

**5. a. Lakukan metode train test split**

In [89]:
df1.customer_type.unique()

array(['-', 'good'], dtype=object)

In [90]:
# Import train_test_split
from sklearn.model_selection import train_test_split

# X merupakan dataset kecuali target variabel
X = df1.drop(['customer_type'], axis=1)

# Y merupakan target variabel yaitu Species
Y = df1.customer_type

# kita gunakan 80% data training
split_size=0.2

# Creation of Train and Test dataset
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size=split_size,random_state=22)

# Creation of Train and validation dataset
X_train, X_val, y_train, y_val = train_test_split(X_train,y_train,test_size=0.2,random_state=5)


Selanjutnya kita akan meng-import beberapa alat pengukuran 

In [91]:
from sklearn import model_selection
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [115]:
# import library
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

# buat objek model
models = []

names = ['LR','Random Forest','Neural Network','GaussianNB','DecisionTreeClassifier','SVM',]
b
models.append((LogisticRegression()))
models.append((RandomForestClassifier(n_estimators=100)))

models.append((MLPClassifier()))
models.append((GaussianNB()))
models.append((DecisionTreeClassifier()))
models.append((SVC()))

# fit model
kfold = model_selection.KFold(n_splits=5,random_state=7)

for i in range(0,len(models)):    
    cv_result = model_selection.cross_val_score(models[i],X_train,y_train,cv=kfold,scoring='accuracy')
    score=models[i].fit(X_train,y_train)
    prediction = models[i].predict(X_val)
    acc_score = accuracy_score(y_val,prediction)     
    print ('-'*40)
    print ('{0}: {1}'.format(names[i],acc_score))
    
randomForest = RandomForestClassifier(n_estimators=100)
randomForest.fit(X_train,y_train)


# panggil fungsi predict
prediction = randomForest.predict(X_test)

/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will 

----------------------------------------
LR: 0.4642857142857143
----------------------------------------
Random Forest: 0.5833333333333334
----------------------------------------
Neural Network: 0.36904761904761907
----------------------------------------
GaussianNB: 0.4880952380952381
----------------------------------------
DecisionTreeClassifier: 0.5714285714285714
----------------------------------------
SVM: 0.36904761904761907


/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/si

In [116]:
print ('-'*40)
print ('Accuracy score:')
print (accuracy_score(y_test, prediction))
print ('-'*40)
print ('Confusion Matrix:')
print (confusion_matrix(y_test, prediction))
print ('-'*40)
print ('Classification Matrix:')
print (classification_report(y_test, prediction))

----------------------------------------
Accuracy score:
0.6666666666666666
----------------------------------------
Confusion Matrix:
[[25 17]
 [18 45]]
----------------------------------------
Classification Matrix:
              precision    recall  f1-score   support

           -       0.58      0.60      0.59        42
        good       0.73      0.71      0.72        63

    accuracy                           0.67       105
   macro avg       0.65      0.65      0.65       105
weighted avg       0.67      0.67      0.67       105



Apabila akurasi model Anda masih **di bawah 80%, maka Lakukan langkah hyperparameter tuning** dengan mengubah konfigurasi default.

In [118]:
from sklearn.model_selection import GridSearchCV
def print_results(results):
    print('BEST PARAMS: {}\n'.format(results.best_params_))

    means = results.cv_results_['mean_test_score']
    stds = results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean, 3), round(std * 2, 3), params))

In [119]:
lr = LogisticRegression()
parameters = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters, cv=5)
cv.fit(X_train, y_train.values.ravel())

print_results(cv)

/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will 

BEST PARAMS: {'C': 0.01}

0.631 (+/-0.086) for {'C': 0.001}
0.661 (+/-0.111) for {'C': 0.01}
0.655 (+/-0.089) for {'C': 0.1}
0.655 (+/-0.084) for {'C': 1}
0.649 (+/-0.097) for {'C': 10}
0.649 (+/-0.097) for {'C': 100}
0.649 (+/-0.097) for {'C': 1000}


/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)
/Users/hargi/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [ ]:
#Function to print best hyperparamaters: 
def print_best_params(gd_model):
    param_dict = gd_model.best_estimator_.get_params()
    model_str = str(gd_model.estimator).split('(')[0]
    print("\n*** {} Best Parameters ***".format(model_str))
    for k in param_dict:
        print("{}: {}".format(k, param_dict[k]))
    print()

    
#test train split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 0 )



**5. b. Jelaskan hasil pada model Anda di atas**<br/>
- Apakah sudah sesuai dengan keinginan dan ekspektasi Anda?
- Model apa yang memiliki tingkat akurasi tertinggi?

In [ ]:
Tidak, hasil permodelan confution matrix dibawah 80%

In [ ]:
Random Forest